Sources/OpenAPIRuntime/Interface/CurrencyTypes.swift

```diff
@@ -11,281 +11,100 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 //===----------------------------------------------------------------------===//
-#if canImport(Darwin)
-import Foundation
-#else
-@preconcurrency import struct Foundation.Data
-@preconcurrency import struct Foundation.URLQueryItem
-#endif
 
-/// A header field used in an HTTP request or response.
-public struct HeaderField: Hashable, Sendable {
+import HTTPTypes
 
-    /// The name of the HTTP header field.
-    public var name: String
+/// A container for request metadata already parsed and validated
+/// by the server transport.
+public struct ServerRequestMetadata: Hashable, Sendable {
 
-    /// The value of the HTTP header field.
-    public var value: String
+    /// The path parameters parsed from the URL of the HTTP request.
+    public var pathParameters: [String: Substring]
 
-    /// Creates a new HTTP header field.
+    /// Creates a new metadata wrapper with the specified path and query parameters.
     /// - Parameters:
-    ///   - name: A name of the HTTP header field.
-    ///   - value: A value of the HTTP header field.
-    public init(name: String, value: String) {
-        self.name = name
-        self.value = value
+    ///   - pathParameters: Path parameters parsed from the URL of the HTTP
+    ///     request.
+    public init(
+        pathParameters: [String: Substring] = [:]
+    ) {
+        self.pathParameters = pathParameters
     }
 }
 
-/// Describes the HTTP method used in an OpenAPI operation.
-///
-/// https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-7
-public struct HTTPMethod: RawRepresentable, Hashable, Sendable {
+extension HTTPRequest {
 
-    /// Describes an HTTP method explicitly supported by OpenAPI.
-    private enum OpenAPIHTTPMethod: String, Hashable, Sendable {
-        case GET
-        case PUT
-        case POST
-        case DELETE
-        case OPTIONS
-        case HEAD
-        case PATCH
-        case TRACE
-    }
-
-    /// The underlying HTTP method.
-    private let value: OpenAPIHTTPMethod
-
-    /// Creates a new method from the provided known supported HTTP method.
-    private init(value: OpenAPIHTTPMethod) {
-        self.value = value
+    /// Creates a new request.
+    /// - Parameters:
+    ///   - path: The URL path of the resource.
+    ///   - method: The HTTP method.
+    ///   - headerFields: The HTTP header fields.
+    public init(soar_path path: String, method: Method, headerFields: HTTPFields = .init()) {
+        self.init(method: method, scheme: nil, authority: nil, path: path, headerFields: headerFields)
     }
 
-    public init?(rawValue: String) {
-        guard let value = OpenAPIHTTPMethod(rawValue: rawValue) else {
+    /// The query substring of the request's path.
+    public var soar_query: Substring? {
```

Comment:
> Are they used by the end-user or the transport?

They're used by the generated code but _also_ by transports and middlewares, as without a method like this, I found that I had to reimplement splitting the query out of the path when adapting existing transports and middlewares. I personally think `HTTPTypes` itself should provide more granular access to just the path (without query) and the query itself, but in the absence of that, I added it here. If it's not in the API, all transports will have to implement their own variant of parsing this out.

> should we consider making a new SPI

That's an interesting idea. It'd certainly make the different audiences of the runtime API more explicit. However I'm a bit worried about the discoverability, as someone implementing a transport might not realize we'd expect them to `@_spi(Transport) import OpenAPIRuntime` to get all the "good" APIs, plus we use the `@_spi(Generated)` to allow ourselves to be more aggressive with API evolution, not holding all the same API guarantees for the SPI as we do for API (as long as we can prove to ourselves we're not breaking anyone's existing generated code).

Here, since the transports would still be user-written code, we'd have to provide the same API stability guarantees as our vanilla public API, so we wouldn't get much benefit from the SPI itself, I think.

Reply:
Since these are extensions on HTTPRequest/HTTPResponse, only users who are writing a transport or a middleware are expected to even instantiate/use these types. Users who only _use_ transports/middlewares won't see these extensions in autocompletion, since they're never working with these raw request/response types directly.

Reply:
So these moved to the `Generated` SPI (cf. any new API specifically for transports). I think that's fine, because we can defer on the decision on how they go into the API once we've got some feedback from the SSWG?

Reply:
Yup, exactly. The extensions used by generated code moved into the `Generated` SPI, and the other extensions got removed, and each transport has to cargo cult the utils. Fine for now, we can always adopt more utils that show up in swift-http-types later.